### PR TITLE
feat(server): resolve duplicates

### PR DESCRIPTION
**Pull request by Phlogi**
## Description

This PR moves duplicate metadata synchronization logic from the frontend to the backend, making it more robust and enabling proper batch operations. This is an improved and refactored version of #13851.

**What's new:**

- **Server-side metadata sync** - Albums, tags, favorites, ratings, descriptions, visibility, and location are now unconditionally synchronized on the backend when resolving duplicates. All metadata from trashed/deleted assets is merged into kept assets automatically — no per-field toggles or settings needed.
- **Batch resolution API** - New `/duplicates/resolve` endpoint handles multiple duplicate groups in a single request
- **Smart keep suggestions** - Server provides suggestions for which duplicates to keep based on file size and EXIF completeness (same logic as before)
- **Stack from duplicates** - New `/duplicates/stack` endpoint to organize duplicates as stacks instead of deleting them
- **Tags survive re-extraction** - Merged tags are written to `asset_exif.tags` and the property is locked, so subsequent metadata re-extraction (sidecar write → extract metadata) does not overwrite the merged tag set
- **Correct exif handling** - Description and location are properly written to exif data (not asset update), and locked columns are handled by the existing `distinctLocked` mechanism without redundant wrappers

**Not included (planned for follow-up):**

- Bulk "resolve all duplicates" with background job processing - this PR focuses on the core resolution API. If well-received, a follow-up PR will add `POST /duplicates/resolve/bulk` with job queue support for processing all duplicate groups in the background.

## How Has This Been Tested?

- [x] Comprehensive E2E tests (70+ test cases covering all endpoints and edge cases)
- [x] Unit tests for metadata merge logic and keep suggestions algorithm
- [x] Tested all sync options (albums, tags, favorites, ratings, description, location, visibility)
- [x] Validated permission checks and error handling
- [x] Manual testing in web UI
- [x] External automatic testing with a python script and actual images

<details><summary><h2>Screenshots (if appropriate)</h2></summary>

image

</details>

## API Changes

**New endpoints:**

- `POST /duplicates/resolve` - Batch resolve with metadata sync and optional trash/delete
- `POST /duplicates/stack` - Create stack from duplicate group assets

**Modified endpoints:**

- `GET /duplicates` - Now includes `suggestedKeepAssetIds` field in response

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if applicable - _OpenAPI specs updated_
- [x] I have no unrelated changes in the PR. - _Only duplicate resolution features_
- [x] I have confirmed that any new dependencies are strictly necessary. - _No new dependencies_
- [x] I have written tests for new code (if applicable) - _70+ E2E tests + unit tests + external python scripted tests_
- [x] I have followed naming conventions/patterns in the surrounding code - _Follows duplicate/asset patterns_
- [x] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc. - _DuplicateService uses repositories_
- [x] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`) - _Only data access_

## Additional external automated python test cases 
<details><summary>Click to see the cases and reasoning</summary>
<p>

<detail> <summary></summary>

# Duplicate Test Cases

This document enumerates every duplicate merge test set. Each set groups multiple assets for user Eve where one
or more assets are marked **keep** and the rest are marked **trash** to exercise merge
behavior. Each table lists the candidate assets and the expected result.
Total expected duplicate sets: 33.

## Favorites (synchronizeFavorites)

Fields tested: favorite

| Set               | Keep           | Trash Candidates            | Expected Result |
| ----------------- | -------------- | --------------------------- | --------------- |
| Fav-01-NoneVsTrue | favorite=false | favorite=true               | favorite=true   |
| Fav-03-AllFalse   | favorite=false | favorite=false              | favorite=false  |
| Fav-04-MultiMixed | favorite=false | favorite=false, true, false | favorite=true   |

## Rating (synchronizeRating)

Fields tested: rating

| Set               | Keep     | Trash Candidates | Expected Result |
| ----------------- | -------- | ---------------- | --------------- |
| Rat-01-LowVsHigh  | rating=1 | rating=4         | rating=4        |
| Rat-03-AllRange   | rating=0 | rating=2, 3, 5   | rating=5        |
| Rat-04-ZeroVsZero | rating=0 | rating=0         | rating=0        |

## Tags (synchronizeTags)

Fields tested: tags

| Set                | Keep                | Trash Candidates                         | Expected Result             |
| ------------------ | ------------------- | ---------------------------------------- | --------------------------- |
| Tag-01-NoneVsOne   | tags=none           | tags=Beach                               | tags=Beach                  |
| Tag-02-KeepHasTags | tags=Travel, Food   | tags=none                                | tags=Travel, Food           |
| Tag-03-Overlap     | tags=Family, Travel | tags=Travel, Sunset                      | tags=Family, Travel, Sunset |
| Tag-04-MultiMixed  | tags=Work           | tags=Work, Party; tags=none; tags=Hiking | tags=Work, Party, Hiking    |

## Description (synchronizeDescription)

Fields tested: description (unique non-empty, trimmed)

| Set                   | Keep                  | Trash Candidates         | Expected Result               |
| --------------------- | --------------------- | ------------------------ | ----------------------------- |
| Desc-01-EmptyVsFilled | desc=""               | desc="A sunset photo"    | desc="A sunset photo"         |
| Desc-02-FilledVsEmpty | desc="Mountain view"  | desc=""                  | desc="Mountain view"          |
| Desc-03-BothFilled    | desc="First"          | desc="Second"            | desc="First\nSecond"          |
| Desc-04-SameValues    | desc="Same caption"   | desc="Same caption"      | desc="Same caption"           |
| Desc-05-FourMixed     | desc="Alpha"          | desc="Beta", "", "Gamma" | desc="Alpha\nBeta\nGamma"     |
| Desc-06-MultiLineTrim | desc="Line A\nLine B" | desc=" Line B \nLine C " | desc="Line A\nLine B\nLine C" |

## Location (synchronizeLocation)

Fields tested: location

| Set                   | Keep            | Trash Candidates       | Expected Result |
| --------------------- | --------------- | ---------------------- | --------------- |
| Loc-01-NoneVsOne      | location=null   | location=NYC           | location=NYC    |
| Loc-02-OneVsNone      | location=Paris  | location=null          | location=Paris  |
| Loc-03-BothSame       | location=London | location=London        | location=London |
| Loc-04-Different      | location=London | location=Tokyo         | location=London |
| Loc-05-ThreeDifferent | location=NYC    | location=Paris, London | location=NYC    |
| Loc-06-MixedWithNull  | location=NYC    | location=null, London  | location=NYC    |
| Loc-07-AllNull        | location=null   | location=null, null    | location=null   |

## Albums (synchronizeAlbums)

Fields tested: album membership

| Set                          | Keep              | Trash Candidates               | Expected Result                 |
| ---------------------------- | ----------------- | ------------------------------ | ------------------------------- |
| Album-01-TrashHasAlbum       | album=none        | album=TestAlbum-A              | albums=TestAlbum-A              |
| Album-02-KeepHasAlbum        | album=TestAlbum-B | album=none                     | albums=TestAlbum-B              |
| Album-03-DifferentAlbums     | album=TestAlbum-C | album=TestAlbum-D              | albums=TestAlbum-C, TestAlbum-D |
| Album-04-MultipleTrashAlbums | album=none        | album=TestAlbum-E, TestAlbum-F | albums=TestAlbum-E, TestAlbum-F |

## Combined (multiple sync options)

Fields tested: favorite, rating, description, location, album

| Set                   | Keep                                          | Trash Candidates                                                                             | Expected Result                                          |
| --------------------- | --------------------------------------------- | -------------------------------------------------------------------------------------------- | -------------------------------------------------------- |
| Combo-01-FavRating    | favorite=false, rating=1                      | favorite=true rating=3; favorite=false rating=5                                              | favorite=true; rating=5                                  |
| Combo-03-AllFields    | favorite=false rating=1 desc="A" location=NYC | favorite=true rating=3 desc="B" location=null; favorite=false rating=5 desc="" location=null | favorite=true; rating=5; desc="A\nB"; location=NYC       |
| Combo-04-AllWithAlbum | favorite=false rating=2 album=TestAlbum-G     | favorite=true rating=4 album=TestAlbum-H                                                     | favorite=true; rating=4; albums=TestAlbum-G, TestAlbum-H |

## Multi-Keep (multiple keepers)

Fields tested: merged metadata applies to every kept asset
Note: Hidden assets are excluded because they do not show up in the duplicate review UI.

| Set                | Keep                                                                                        | Trash Candidates                         | Expected Result                                                 |
| ------------------ | ------------------------------------------------------------------------------------------- | ---------------------------------------- | --------------------------------------------------------------- |
| Keep-01-FavRating  | favorite=false rating=1; favorite=true rating=2                                             | favorite=false rating=5                  | favorite=true; rating=5                                         |
| Keep-02-TagsAlbums | tags=Work album=TestAlbum-MK-A; tags=Travel album=none                                      | tags=Party album=TestAlbum-MK-B          | tags=Work, Travel, Party; albums=TestAlbum-MK-A, TestAlbum-MK-B |
| Keep-03-VisDescLoc | visibility=timeline desc="Alpha" location=NYC; visibility=archive desc="Beta" location=null | visibility=archive desc="" location=null | visibility=archive; desc="Alpha\nBeta"; location=NYC            |

# Screenshots
image
image
image

</p>
</details> 

## Please describe to which degree, if any, an LLM was used in creating this pull request.

Claude Code assisted with refactoring the sync logic to the backend, writing test cases, and cleaning up the frontend code. All logic was manually reviewed and tested.

### CODE REVIEW

File: e2e/src/utils.ts [hunk 1]
```diff
@@ -499,6 +499,15 @@ export const utils = {
   createStack: (accessToken: string, assetIds: string[]) =>
     createStack({ stackCreateDto: { assetIds } }, { headers: asBearerAuth(accessToken) }),
 
+  setAssetDuplicateId: async (accessToken: string, assetId: string, duplicateId: string | null) => {
```
Comment (jrasm91) [inline e2e/src/utils.ts]: You can set duplicateId via `updateAssets`, so there is no need to do it via the database.

File: docs/docs/features/duplicates-utility.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+| Album       | `false` | When enabled, the kept assets will be added to _every_ album that the other assets in the group belong to, including assets that are selected to be trashed. |
+| Favorite    | `false` | When enabled, if any of the assets in the group have been added to favorites, every kept asset will also be added to favorites.                              |
+| Rating      | `false` | When enabled, if one or more assets in the duplicate group have a rating, the highest rating is selected and then synchronized to the kept assets.           |
+| Description | `false` | When enabled, descriptions from each asset are combined together and then synchronized to all the kept assets.                                               |
+| Visibility  | `false` | When enabled, the most restrictive visibility is applied the the kept assets.                                                                                |
+| Location    | `false` | When enabled, latitude and longitude are only copied if among the group there is a single asset with geolocation data.                                       |
```
Comment (jrasm91) [inline docs/docs/features/duplicates-utility.md]: Do you mind adding an item for tags here?

File: web/src/lib/stores/preferences.store.ts [hunk 3]
```diff
@@ -149,3 +151,17 @@ export const autoPlayVideo = persisted<boolean>('auto-play-video', true, {});
...
+export const duplicateSettings = new PersistedLocalStorage<DuplicateSettings>(
+  'duplicate-settings',
+  {
+    syncAlbums: false,
+    syncVisibility: false,
+    syncFavorites: false,
+    syncRating: false,
+    syncDescription: false,
+    syncLocation: false,
+    syncTags: false,
```
Comment (mertalev) [inline web/src/lib/stores/preferences.store.ts]: Do you think these need to be configurable? Would anyone actually not want to sync this metadata?

Comment (jrasm91) [reply web/src/lib/stores/preferences.store.ts]: Yeah that's a fair point. I do think this would be a good default to have on all the time. Maybe we remove it and then see if anyone asks for it.

Comment (Phlogi) [reply web/src/lib/stores/preferences.store.ts]: As I tried to prove in detail,the logic is working, there is no reason to have it disabled. You'd only lose metadata info. So I support changing these.

Comment (jrasm91) [reply web/src/lib/stores/preferences.store.ts]: Feel free to delete the query parameters and update the code to always sync everything.

Comment (Phlogi) [reply web/src/lib/stores/preferences.store.ts]: I would keep them in the server endpoint for external tools e.g. mass de-duplication where a user wants to not sync certain metadata on purpose. Maybe overthinking it though.

File: e2e/src/utils.ts (hunk 1 above)
Comment (Phlogi) [reply e2e/src/utils.ts]: https://github.com/immich-app/immich/pull/25316/commits/74402ea99f2fa2b3b98b1d7551fa20ff570f9d83

File: web/src/lib/stores/preferences.store.ts (hunk 3 above)
Comment (jrasm91) [reply web/src/lib/stores/preferences.store.ts]: Yes, please remove them.

**Comment by Phlogi** [comment]
@jrasm91 
Significant amount of the functional tests were showing the broken logic. 
I restored it, the sync of tags was particularly difficult.

File: web/src/lib/stores/preferences.store.ts (hunk 3 above)
Comment (Phlogi) [reply web/src/lib/stores/preferences.store.ts]: will do as next and is the remaining 
open thing

**Comment by Phlogi** [comment]
@jrasm91 Done, the PR description was updated too. Please review. 
All tests run successful on my side. 
Without the settings now, how/where do we link to documentation?

**Comment by jrasm91** [comment]
Why don't we just stick it here for now?

image


So maybe just like:
> Resolve each group by indicating which, if any, are duplicates. For more information, refer to the documentation.

I see the modal shows some information about the automatic deduping, but I think we should probably delete that and migrate that content to the documentation page as well. I'm fine if you want to do that in this PR as well.

File: docs/docs/features/duplicates-utility.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+| Favorite    | `false` | When enabled, if any of the assets in the group have been added to favorites, every kept asset will also be added to favorites.                    |
+| Rating      | `false` | When enabled, if one or more assets in the duplicate group have a rating, the highest rating is selected and then synchronized to the kept assets. |
+| Description | `false` | When enabled, descriptions from each asset are combined together and then synchronized to all the kept assets.                                     |
+| Visibility  | `false` | When enabled, the most restrictive visibility is applied the the kept assets.                                                                      |
+| Location    | `false` | When enabled, latitude and longitude are only copied if all assets with geolocation data in the group share the same coordinates.                  |
+| Tag         | `false` | When enabled, tags from all assets in the group are merged and applied to every kept asset.                                                        |
```
Comment (NicholasFlamy) [inline docs/docs/features/duplicates-utility.md]: Did y'all make it Default to true? If so this needs to be updated.
```suggestion
| Name        | Default | Description                                                                                                                                        |
| ----------- | ------- | -------------------------------------------------------------------------------------------------------------------------------------------------- |
| Album       | `true` | When enabled, the kept assets will be added to _every_ album that the other assets in the group belong to.                                         |
| Favorite    | `true` | When enabled, if any of the assets in the group have been added to favorites, every kept asset will also be added to favorites.                    |
| Rating      | `true` | When enabled, if one or more assets in the duplicate group have a rating, the highest rating is selected and then synchronized to the kept assets. |
| Description | `true` | When enabled, descriptions from each asset are combined together and then synchronized to all the kept assets.                                     |
| Visibility  | `true` | When enabled, the most restrictive visibility is applied the the kept assets.                                                                      |
| Location    | `true` | When enabled, latitude and longitude are only copied if all assets with geolocation data in the group share the same coordinates.                  |
| Tag         | `true` | When enabled, tags from all assets in the group are merged and applied to every kept asset.                                                        |
```

Comment (Phlogi) [reply docs/docs/features/duplicates-utility.md]: Done.

**Comment by Phlogi** [comment]
> Why don't we just stick it here for now?
> image
> 
> So maybe just like:
> 
> > Resolve each group by indicating which, if any, are duplicates. For more information, refer to the documentation.
> 
> I see the modal shows some information about the automatic deduping, but I think we should probably delete that and migrate that content to the documentation page as well. I'm fine if you want to do that in this PR as well.

@jrasm91 Not sure how to properly link to it. Could you maybe implement that part?

**Comment by jrasm91** [comment]
Yeah I can do it.

**Comment by Phlogi** [comment]
> Yeah I can do it.

Please review f899c7ad30a9d55c88eb02d64b3a839a68bf1d34.

**Comment by Phlogi** [comment]
@NicholasFlamy Are you fine with the changes?

**Comment by NicholasFlamy** [comment]
> @NicholasFlamy Are you fine with the changes?

Looks decent, both the docs and the tool. Now you need approval from the team.

**Review by jrasm91 — APPROVED** [review]

**Comment by jrasm91** [comment]
@Phlogi I think this is ready to go in the next release. If you have time, do you mind doing another round of testing to make sure it still works correctly?

**Comment by Phlogi** [comment]
> @Phlogi I think this is ready to go in the next release. If you have time, do you mind doing another round of testing to make sure it still works correctly?

Great, ran my test sets and it looks good. One thing I noticed about the rating value: `expected=0 actual=None`.

File: server/src/services/duplicate.service.ts (lines 256-263)

Small issue: The merge always sets exifUpdate.rating = 0, even when no duplicate has a rating. 
This writes 0 to asset_exif.rating, triggers a SidecarWrite job, and then MetadataExtraction normalizes it back to null — a no-op round-trip. It also means exifUpdate is never empty, so SidecarWrite fires on every merge even when there's nothing meaningful to persist.

Fix: Guard the rating assignment the same way description and latitude/longitude are guarded right below it:

```
if (rating > 0) {
  response.exifUpdate.rating = rating;
}
```

@jrasm91 it's almost cosmetic, but could be implemented, let me know.

**Comment by jrasm91** [comment]
Yeah feel free to push that change. Thanks for testing!

**Comment by Phlogi** [comment]
> Yeah feel free to push that change. Thanks for testing!

Done, also added a missing type cast.
